calendar/jwt_oauth2_handler.bal

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/http;
+import ballerina/jwt;
+import ballerina/log;
+
+const string OAUTH_URL = "https://oauth2.googleapis.com";
+type Error distinct error;
+
+# Defines the OAuth2 handler for client authentication.
+public client class ClientOAuth2ExtensionGrantHandler {
+
+    OAuth2ExtensionGrantProvider provider;
+
+    public isolated function init(jwt:IssuerConfig? config = ()) {
+        self.provider = new(config);
+    }
+
+    # Returns the headers map with the relevant authentication requirements.
+    # 
+    # + userAccount - The email address of the user for requesting delegated access in service account
+    # + return - The updated headers map or else an `Error` in case of an error
+    public isolated function getSecurityHeaders(string userAccount) returns map<string>|Error {
+        string|Error result = self.provider.generateToken(userAccount);
+        if (result is string) {
+            map<string> headers = {};
+            headers[http:AUTH_HEADER] = http:AUTH_SCHEME_BEARER + " " + result;
+            return headers;
+        } else {
+            return prepareError("Failed to enrich headers with OAuth2 token.", result);
+        }
+    }
+}
+
+public class OAuth2ExtensionGrantProvider {
+
+    jwt:IssuerConfig jwtIssuerConfig = {};
+    boolean isServiceAccount = false;
+    http:Client clientEndpoint;
+    
+    public isolated function init(jwt:IssuerConfig? config) {
+        self.clientEndpoint = checkpanic new(OAUTH_URL);
```

Comment:
why do we need `checkpanic` here?

Reply:
Here you are trying to init the `http:Client` for the given URL. What if we return the error in case of failure? As I know, we must avoid using `checkpanic` as much as possible.

Reply:
Yes. We have to avoid. But here, if we use `check` and return the error, we have to use `checkpanic` in https://github.com/Chuhaa/module-ballerinax-googleapis.calendar/blob/9cebe3b0975ef9172cc820356361af470f57194a/calendar/endpoints.bal#L27

Reply:
No right? You can do `ClientOAuth2ExtensionGrantHandler` initialization in the `init` and handle the error.

Reply:
Yes. Moved that initialization into https://github.com/Chuhaa/module-ballerinax-googleapis.calendar/blob/4bb9d3a406e1609071b2b2707cbaa606fa62643f/calendar/endpoints.bal#L36